File: STACMR-R/staMR.R
staMR <- function(data=list(), partial = list(), shrink=-1) {
  # function [xPrime, fit, shrinkage] = staMR (data, partial, shrink)
  # fits monotonic regression model to data according to partial order
  # data is list of lists of data or structured output from staSTATS
  # partial is partial order in list format
  # shrink is parameter to control shrinkage of covariance matrix;
  # 0 = no shrinkage; 1 = diagonal matrix; -1 = calculate optimum
  # returns:
  #   x = best fitting MR values to y-means
  #   f = total fit statistic
  #   shrinkage = shrinkage from applying staSTATS
  # *************************************************************************
  # modified from matlab 13 September 2016
  # *************************************************************************
  #

  # set up defaults
  tol <- 10e-5
  if (missing(partial)) {partial = list()}
  if (missing(shrink)) {shrink = -1}
  
  # get stats from data (depending on its form)
  if (is(data,"data.frame")) {
    y = gen2list (data) # convert from general format
    y = staSTATS (y, shrink) # get stats
  } else if (is.null(data[[1]]$means)) {y = staSTATS(data, shrink) # in list form, get stats
  } else {y = data} # already in stats form
  
  # convert partial order to list if in adjacency matrix form
  if (is(partial,"matrix")) {partial = adj2list(partial)}
  
  # extract shrinkage parameters (for information only)
  nvar = length(y)
  shrinkage = matrix(0, length(y[[1]]$shrinkage), nvar)
  for (ivar in 1:nvar) {shrinkage[,ivar] = y[[ivar]]$shrinkage}

  # do MR for each dependent variable
  xPrime = vector("list", nvar)
  fit = matrix(0, nvar, 1)
  for (ivar in 1:nvar) {
    out = jMR (y[[ivar]]$means, y[[ivar]]$weights, partial)
    xPrime[[ivar]] = out$x
    fit[ivar] = out$fval
  }
  fval = sum(fit)
  if (fval < tol) {fval = 0} # round down
  
  for (i in 1:nvar) {xPrime[[i]]=matrix(xPrime[[i]],length(xPrime[[i]]),1)}
  output = list(xPrime, fval, shrinkage)
  names(output) = c("x", "fval", "shrinkage")
  
  return(output)
}

File: STACMR-matlab/staMR.m
function [xPrime, fit, shrinkage] = staMR (data, E, shrink, reverse)
% function [xPrime, fit, shrinkage] = staMR (data, E, shrink, reverse)
% fits monotonic regression model to data according to partial order
% data is cell array of data or structured output from staSTATS
% E is partial order
% shrink is parameter to control shrinkage of covariance matrix;
% 0 = no shrinkage; 1 = diagonal matrix; -1 = calculate optimum
% returns:
% x = best fitting MR values to y-means
% f = total fit statistic
% *************************************************************************
% Last updated: 4 Sept 2015
% modified 3 June 2016 to deal with cell array data format
% *************************************************************************
%
tol = 1e-10;
if nargin < 4, reverse = false; end
if nargin < 3, shrink = -1; end
if nargin < 2, E = {}; end

if isempty(E), E = {}; end
if ~iscell(E), E = {E}; end
if isempty(shrink), shrink = -1; end
if isempty(reverse), reverse = 0; end

if iscell(data)
    if isstruct(data{1})
        y = data; % if structured then already in stats form
    else
        y = staSTATS(data, shrink); % otherwise assume within-subjects data and get stats
    end
else
    if isstruct(data) 
        y = cell(1,1); y{1} = data; % data is structured output for one variable
    else
        celldata = gen2cell (data); % convert from old "general format"
        y = staSTATS (celldata, shrink);
    end
end

nvar = numel(y);
shrinkage = zeros(numel(y{1}.shrinkage),nvar); 
for ivar=1:nvar, shrinkage(:,ivar) = y{ivar}.shrinkage; end

xPrime = cell(1,nvar); 
fits = zeros(nvar,1); 
for ivar = 1:nvar
    [xPrime{ivar}, fits(ivar)] = jMR (y{ivar}.means, y{ivar}.weights, E, reverse);
end
fit = sum(fits);
if fit < tol, fit=0; end



